### ISSUE TITLE
Custom webpack config file makes the prod build fail

### ISSUE THREAD

**Issue by rolandjitsu** [driftyco/ionic-app-scripts#393]
**Note: for support questions, please use one of these channels:**

https://forum.ionicframework.com/
http://ionicworldwide.herokuapp.com/

#### Short description of the problem:
I seems that if I use
```js
"config": {
	"ionic_webpack": "./config/webpack.config.js"
}
```
The prod builds fail with:
```shell
➜  operator git:(master) ✗ npm run ionic:build

> @infarm/operator@0.0.0 ionic:build /Users/rolandjitsu/Projects/infarm/operator
> ionic-app-scripts build

[12:03:17]  ionic-app-scripts 0.0.43 
[12:03:18]  build prod started ... 
[12:03:18]  clean started ... 
[12:03:18]  clean finished in 4 ms 
[12:03:18]  copy started ... 
[12:03:18]  ngc started ... 
[12:03:18]  copy finished in 138 ms 
[12:03:39]  ngc finished in 21.04 s 
[12:03:39]  webpack started ... 
[12:03:39]  build prod failed: Cannot read property 'type' of undefined 
[12:03:39]  ionic-app-script task: "build" 
[12:03:39]  TypeError: Cannot read property 'type' of undefined 
```

#### What behavior are you expecting?
To build properly without errors.

**Steps to reproduce:**
1. Run `npm run ionic:build` with the custom config file provided bellow

```
// For reference, check the following:
// https://github.com/driftyco/ionic-app-scripts/blob/master/config/webpack.config.js
// Config borrowed from: https://github.com/driftyco/ionic-cli/issues/1205#issuecomment-257634690
const path = require('path');
const webpack = require('webpack');

const ionicWebpackFactoryPath = path.join(process.env.IONIC_APP_SCRIPTS_DIR, 'dist', 'webpack', 'ionic-webpack-factory.js');
const ionicWebpackFactory = require(ionicWebpackFactoryPath);


function getEntryPoint() {
	if (process.env.IONIC_ENV === 'prod') {
		return '{{TMP}}/app/main.prod.js';
	}
	return '{{SRC}}/app/main.dev.ts';
}

function getDevtool() {
	if (process.env.IONIC_ENV === 'prod') {
		// For now, just force source-map for prod builds.
		return 'source-map';
	}
	return process.env.IONIC_SOURCE_MAP;
}

function getSourcemapLoader() {
	if (process.env.IONIC_ENV === 'prod') {
		// TODO figure out the disk loader, it's not working yet
		return [];
	}

	return [
		{
			test: /\.ts$/,
			loader: path.join(process.env.IONIC_APP_SCRIPTS_DIR, 'dist', 'webpack', 'typescript-sourcemap-loader-memory.js')
		}
	];
}

function getPlugins() {
	const plugins = [
		new webpack.DefinePlugin({
			'IONIC_ENV': JSON.stringify(process.env.IONIC_ENV)
		})
	];

	if (process.env.IONIC_ENV === 'prod') {
		return plugins.push(
			// This helps ensure the builds are consistent if source hasn't changed:
			new webpack.optimize.OccurrenceOrderPlugin(),

			// Try to dedupe duplicated modules, if any:
			new webpack.optimize.DedupePlugin()
		);
	} else {
		// For dev builds, use Ionic custom environment.
		plugins.push(ionicWebpackFactory.getIonicEnvironmentPlugin());
	}

	return plugins;
}


module.exports = {
	entry: getEntryPoint(),
	output: {
		path: '{{BUILD}}',
		filename: 'main.js'
	},
	devtool: getDevtool(),
	resolve: {
		extensions: ['.js', '.ts', '.json']
	},
	module: {
		loaders: [
			{
				test: /\.json$/,
				loader: 'json'
			}
		].concat(getSourcemapLoader())
	},
	plugins: getPlugins(),
	node: {
		fs: 'empty',
		net: 'empty',
		tls: 'empty'
	}
};
```

**Which @ionic/app-scripts version are you using?**
`0.0.43`

**Other information:** (e.g. stacktraces, related issues, suggestions how to fix, stackoverflow links, forum links, etc)
Full npm error log:
```
0 info it worked if it ends with ok
1 verbose cli [ '/Users/rolandjitsu/.nvm/versions/node/v6.7.0/bin/node',
1 verbose cli   '/Users/rolandjitsu/.nvm/versions/node/v6.7.0/bin/npm',
1 verbose cli   'run',
1 verbose cli   'ionic:build' ]
2 info using npm@3.10.3
3 info using node@v6.7.0
4 verbose run-script [ 'preionic:build', 'ionic:build', 'postionic:build' ]
5 info lifecycle @infarm/operator@0.0.0~preionic:build: @infarm/operator@0.0.0
6 silly lifecycle @infarm/operator@0.0.0~preionic:build: no script for preionic:build, continuing
7 info lifecycle @infarm/operator@0.0.0~ionic:build: @infarm/operator@0.0.0
8 verbose lifecycle @infarm/operator@0.0.0~ionic:build: unsafe-perm in lifecycle true
9 verbose lifecycle @infarm/operator@0.0.0~ionic:build: PATH: /Users/rolandjitsu/.nvm/versions/node/v6.7.0/lib/node_modules/npm/bin/node-gyp-bin:/Users/rolandjitsu/Projects/infarm/operator/node_modules/.bin:/Users/rolandjitsu/.nvm/versions/node/v6.7.0/bin:/Users/rolandjitsu/.rvm/gems/ruby-2.1.2/bin:/Users/rolandjitsu/.rvm/gems/ruby-2.1.2@global/bin:/Users/rolandjitsu/.rvm/rubies/ruby-2.1.2/bin:/Users/rolandjitsu/.cabal/bin:/Users/rolandjitsu/.nvm/versions/node/v6.7.0/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Users/rolandjitsu/.rvm/bin:/usr/local/opt/go/libexec/bin:/Users/rolandjitsu/Projects/go/bin:/Users/rolandjitsu/Projects/infarm/operator/node_modules/.bin
10 verbose lifecycle @infarm/operator@0.0.0~ionic:build: CWD: /Users/rolandjitsu/Projects/infarm/operator
11 silly lifecycle @infarm/operator@0.0.0~ionic:build: Args: [ '-c', 'ionic-app-scripts build' ]
12 silly lifecycle @infarm/operator@0.0.0~ionic:build: Returned: code: 1  signal: null
13 info lifecycle @infarm/operator@0.0.0~ionic:build: Failed to exec ionic:build script
14 verbose stack Error: @infarm/operator@0.0.0 ionic:build: `ionic-app-scripts build`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (/Users/rolandjitsu/.nvm/versions/node/v6.7.0/lib/node_modules/npm/lib/utils/lifecycle.js:242:16)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at EventEmitter.emit (events.js:191:7)
14 verbose stack     at ChildProcess.<anonymous> (/Users/rolandjitsu/.nvm/versions/node/v6.7.0/lib/node_modules/npm/lib/utils/spawn.js:40:14)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at ChildProcess.emit (events.js:191:7)
14 verbose stack     at maybeClose (internal/child_process.js:877:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
15 verbose pkgid @infarm/operator@0.0.0
16 verbose cwd /Users/rolandjitsu/Projects/infarm/operator
17 error Darwin 16.1.0
18 error argv "/Users/rolandjitsu/.nvm/versions/node/v6.7.0/bin/node" "/Users/rolandjitsu/.nvm/versions/node/v6.7.0/bin/npm" "run" "ionic:build"
19 error node v6.7.0
20 error npm  v3.10.3
21 error code ELIFECYCLE
22 error @infarm/operator@0.0.0 ionic:build: `ionic-app-scripts build`
22 error Exit status 1
23 error Failed at the @infarm/operator@0.0.0 ionic:build script 'ionic-app-scripts build'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the @infarm/operator package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     ionic-app-scripts build
23 error You can get information on how to open an issue for this project with:
23 error     npm bugs @infarm/operator
23 error Or if that isn't available, you can get their info via:
23 error     npm owner ls @infarm/operator
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

**Comment by rolandjitsu** [comment]
@danbucholtz any clue what is going on here? I cannot find any `.type` in any of the code I wrote, nor in the libraries I wrote and use (not sure about the rest of the libs though). And it only happens with a custom webpack config file added to `package.json` (I need that in order to provide `IONIC_ENV` in my app).

**Comment by danbucholtz** [comment]
I'll resolve this shortly.

Thanks,
Dan

**Comment by rolandjitsu** [comment]
@danbucholtz on this note, do you have any clue how to solve the problem with env variables? 

Because the prod builds seem to loose the values of the env variables due to AOT (I am using a solution similar to [#204](https://github.com/driftyco/ionic-app-scripts/pull/204) for providing the necessary values). And I'm in dire need of a solution so I can run my app on devices and be able to make calls to a different API endpoint than the one I use in dev mode.

**Comment by marcoturi** [comment]
@rolandjitsu meanwhile if you need to inject vars, take a look [here](https://github.com/marcoturi/ionic2-boilerplate/blob/master/config/webpack.config.js).

**Comment by rolandjitsu** [comment]
@marcoturi that does not help :)

**Comment by ivanmayes** [comment]
Having a similar issue, let me know if its different and I can set a different ticket for it:

Here is my webpack.config.js:
```
// Overwrite some of the vars for the Ionic webpack build
var path = require('path');
var webpack = require('webpack');

module.exports = {
  resolve: {
    extensions: ['.js', '.ts', '.json'],
    modules: [path.resolve('node_modules')],
    alias: {
      'app': path.resolve('src/app'),
	  'project': path.resolve('src/project')
    }
  }
};
```

What i'm really trying to accomplish is adding the aliases so that we can use them in our build aka `import { ProductConfig } from 'project/config'`.  

This works great along with adding the paths to our tsconfig.json for `ionic serve`, however the build function isn't picking up the aliases and therefore isn't seeing the imports.

**Comment by danbucholtz** [comment]
@rolandjitsu,

You could try adding node typings to your project. I'm not sure if that will work or not. I have an API in place for doing things like text replacement but it is still being reviewed by the team.

@marcoturi, nice work as always.

@ivanmayes,

I really don't recommend doing that. It would be very challenging to support on our end.

Thanks,
Dan

**Comment by rolandjitsu** [comment]
@danbucholtz Unfortunately I already had `"@types/node": "^6.0.46"` installed and no luck. But I can wait until the feature is released, I'm using rollup for now.

**Comment by ivanmayes** [comment]
@danbucholtz 

Its challenging to support the webpack config? https://webpack.github.io/docs/configuration.html#resolve-alias

I guess I'm not understanding why it's working on ionic serve, but not working on ionic build with my custom webpack.config, seems like build should honor it the same as serve is currently doing.

**Comment by danbucholtz** [comment]
@ivanmayes,

Production builds via `ngc` work differently than development builds. I really don't recommend tinkering with those settings yet. In the future we will likely get mature enough to support them, but I almost guarantee it will be more trouble than it's worth right now.

Thanks,
Dan

**Comment by danbucholtz** [comment]
Hi all,

This is resolved in `0.0.48` which was just published.

Please let me know if you have any issues.

Thanks,
Dan
